src/game_logic.rs

```diff
@@ -227,14 +154,20 @@ impl GameplayState {
     }
 
     pub fn new_with_ui(mut board: BoardState) -> Box<TakeTurnState> {
-        let gameplay_state = Box::new(Self::new(board));
+        let mut gameplay_state = Box::new(Self::new(board));
         println!("Wrapping this gameplay state: {:?}", gameplay_state);
-        TakeTurnState::new(gameplay_state)
+        gameplay_state.take_turn()
     }
 
     pub fn get_board(&self) -> &BoardState {
         &self.board
     }
+
+    // Performs all operations needed before switching to TakeTurnState
+    fn take_turn(&mut self) -> Box<TakeTurnState> {
+        self.board.update_availability();
+        TakeTurnState::new(Box::new(take(self)))
+    }
 }
 
```

Comment:
I am not completely happy with this solution, but I needed to proc this code for all switches to TakeTurnState. This is IMHO better that call update_availability in each relevant branch of code (there are 3+)